src/mesh/tdycoremesh.c

```diff
@@ -489,6 +492,52 @@ PetscErrorCode SaveMeshGeometricAttributes(TDy tdy) {
     ierr = VecDestroy(&natural); CHKERRQ(ierr);
     ierr = VecDestroy(&global); CHKERRQ(ierr);
     ierr = VecDestroy(&local); CHKERRQ(ierr);
+
+    char file[PETSC_MAX_PATH_LEN];
+    PetscBool opt;
+    ierr = PetscOptionsGetString(NULL,NULL,"-tdy_connected_region",file,sizeof(file),
+                               &opt); CHKERRQ(ierr);
+    if (opt) {
+
+      PetscViewer viewer;
+      Vec region_id_nat_idx;
+
+      ierr = VecCreate(PETSC_COMM_WORLD,&region_id_nat_idx); CHKERRQ(ierr);
+
+      ierr = PetscViewerBinaryOpen(PETSC_COMM_WORLD,file,FILE_MODE_READ,&viewer); CHKERRQ(ierr);
+      ierr = VecLoad(region_id_nat_idx,viewer); CHKERRQ(ierr);
+      ierr = PetscViewerDestroy(&viewer); CHKERRQ(ierr);
+
+      // Map natural IDs in global order
+      ierr = DMCreateGlobalVector(dm, &global);CHKERRQ(ierr);
+      ierr = DMPlexNaturalToGlobalBegin(dm, region_id_nat_idx, global);CHKERRQ(ierr);
+      ierr = DMPlexNaturalToGlobalEnd(dm, region_id_nat_idx, global);CHKERRQ(ierr);
+
+      // Map natural IDs in local order
+      ierr = DMCreateLocalVector(dm, &local);CHKERRQ(ierr);
+      ierr = DMGlobalToLocalBegin(dm, global, INSERT_VALUES, local); CHKERRQ(ierr);
+      ierr = DMGlobalToLocalEnd(dm, global, INSERT_VALUES, local); CHKERRQ(ierr);
+
+      // Save the region ids
+      ierr = VecGetLocalSize(local, &size_local);
+      ierr = VecGetArray(local, &p); CHKERRQ(ierr);
+
+      PetscInt ncells = size_local/num_fields;
+      PetscInt *cell_ids;
```

Comment:
Since we're using C and not C++, we can actually take advantage of variable length arrays (VLAs) here. It avoids some extra allocations and is actually easier to understand.

If you want to do this, you can replace lines 526-527 with

```
PetscInt cell_ids[ncells];  // Notice that ncells is a variable, not a constant!
```

This allocates a VLA on the stack of length `ncells`. You can use it just like you'd use any other array. Then just delete line 531.

This isn't possible in C++, so folks who think that C is a proper subset of C++ aren't aware of this trick.

Reply:
I have made the change